src/components/smartSearch/EditTargetDialog.tsx

```diff
@@ -0,0 +1,128 @@
+import { useState } from 'react';
+import { Box, Button, ButtonBase, Card, CardContent, Dialog, DialogContent, Typography } from '@material-ui/core';
+
+import All from './filters/All';
+import Filter from './Filter';
+import { isFilterWithId } from './utils';
+import MostActive from './filters/MostActive';
+import patchTaskTarget from 'fetching/tasks/patchTaskTarget';
+import { useRouter } from 'next/router';
+import { ZetkinSmartSearchFilter } from 'types/zetkin';
+import { FILTER_TYPE, ZetkinSmartSearchFilterWithId } from 'types/smartSearch';
+import { useMutation, useQueryClient } from 'react-query';
+
+interface EditTargetDialogProps {
+    filterSpec: ZetkinSmartSearchFilter[];
+    onDialogClose: () => void;
+    open: boolean;
+}
+
+const EditTargetDialog = ({ onDialogClose, open, filterSpec }: EditTargetDialogProps) : JSX.Element => {
+    const queryClient = useQueryClient();
+    const { orgId, taskId } = useRouter().query;
+    const filtersWithIds = filterSpec.map((filter, index) => ({ ...filter, id: index }));
+    const [filterArray, setFilterArray] = useState<ZetkinSmartSearchFilterWithId[]>(filtersWithIds);
+    const [selectedType, setSelectedType] = useState<FILTER_TYPE | null>(null);
+    const [selectedFilter, setSelectedFilter] = useState<ZetkinSmartSearchFilterWithId| null>(null);
+
+    const taskMutation = useMutation(patchTaskTarget(orgId as string, taskId as string),{
+        onSettled: () => queryClient.invalidateQueries(['filter_spec', orgId, taskId]),
+    } );
+
+    const handleDialogClose = () => {
+        setSelectedType(null);
+        onDialogClose();
+    };
+
+    const handleCancelFilter = () => setSelectedType(null);
+
+    const handleSubmitFilter = (filter:ZetkinSmartSearchFilterWithId | ZetkinSmartSearchFilter) => {
+        if (isFilterWithId(filter)) {
+            setFilterArray(filterArray.map(f => {
+                if (f.id === filter.id) {
+                    return filter;
+                }
+                else {
+                    return f;
+                }
+            }));
+        }
+        else {
+            const newId = 1 + (filterArray.length ? Math.max(...filterArray.map(f => f.id)) : 0);
+            setFilterArray(filterArray.concat({ ...filter, id: newId }));
+        }
+        setSelectedFilter(null);
+        setSelectedType(null);
+    };
+
+    const handleSelectedTypeChange = (type: FILTER_TYPE) => setSelectedType(type);
+
+    const handleSave = () => {
+        taskMutation.mutate(filterArray.map(filter => ({
+            config: filter.config, op: filter.op, type: filter.type,
+        })));
+        onDialogClose();
+    };
+
+    const handleDeleteButtonClick = (filter: ZetkinSmartSearchFilterWithId) => {
+        setFilterArray(filterArray.filter(f => f.id !== filter.id));
+    };
+
+    const handleEditButtonClick = (filter: ZetkinSmartSearchFilterWithId) => {
+        setSelectedFilter(filter);
+        setSelectedType(filter.type as FILTER_TYPE);
```

Comment:
Here we have derived state again. We have the "real" state being set in `setSelectedFilter` and the derived state being set by `setSelectedType()`. What happens if they get out of sync, i.e. if anywhere we forget to set one of them? This is error prone.

I think instead we should just `setSelectedFilter()` and use the `selectedFilter.type` wherever you're currently using `selectedType`.

I realize this can be weird when you're creating a new filter, but there are ways around that (see comment on line 59 above).

Reply:
addressed in #310